src/go/rpk/pkg/cli/cmd/root.go

```diff
@@ -165,6 +201,324 @@ func tryExecPlugin(h pluginHandler, args []string) (string, error) {
+// if a plugin filepath is /foo/bar/rpk.ac-baz_boz-biz, then the pieces will be
+// baz-boz biz.
+type plugin struct {
+	pieces []string
+	path   string
+}
+
+// listPlugins returns all plugins found in fs across the given search
+// directories.
+//
+// Unlike kubectl, we do not allow plugins to be tacked on paths within other
+// plugins. That is, we do not allow rpk-foo-bar to be an additional plugin on
+// top of rpk-foo.
+//
+// We do support plugins defining themselves as "rpk-foo-bar", even though that
+// reserves the "foo" plugin namespace.
+func listPlugins(fs afero.Fs, searchDirs []string) []plugin {
+	searchDirs = uniqueTrimmedStrs(searchDirs)
+
+	uniquePlugins := make(map[string]int) // plugin name (e.g., mm3 or cloud) => index into plugins
+	var plugins []plugin
+	for _, searchDir := range searchDirs {
+		infos, err := afero.ReadDir(fs, searchDir)
+		if err != nil {
+			log.Debugf("unable to read dir %s from PATH: %v", searchDir, err)
+			continue
+		}
+		for _, info := range infos {
+			if info.IsDir() {
+				continue
+			}
+
+			name := info.Name()
+			if !strings.HasPrefix(name, pluginPrefix) && !strings.HasPrefix(name, pluginPrefixAutoComplete) {
+				continue
+			}
+
+			fullPath := filepath.Join(searchDir, name)
+
+			if info.Mode()&0111 == 0 {
+				log.Debugf("matching plugin path %s is not executable, skipping", fullPath)
+				continue
+			}
+
+			// Reverse our args-to-plugin-binary logic to get the
+			// "pieces" that are used to call this plugin.
+			if strings.HasPrefix(name, pluginPrefix) {
+				name = strings.TrimPrefix(name, pluginPrefix)
+			} else {
+				name = strings.TrimPrefix(name, pluginPrefixAutoComplete)
+			}
+			if len(name) == 0 { // e.g., "rpk-"
+				log.Debugf("invalid empty plugin name at path %s", fullPath)
+				continue
+			}
+
+			pieces := toPieces(name)
+			pluginName := pieces[0]
+
+			if priorAt, exists := uniquePlugins[pluginName]; exists {
+				log.Debugf("skipping duplicate plugin %s at path %s, because it is previously overshadowed by path %s", pluginName, fullPath, plugins[priorAt].path)
+				continue
+			}
+
+			uniquePlugins[name] = len(plugins)
+			plugins = append(plugins, plugin{
+				pieces: pieces,
+				path:   fullPath,
+			})
+
+		}
+	}
+
+	return plugins
+}
+
+// Converts a command name to its argument pieces.
+func toPieces(command string) []string {
+	pieces := strings.Split(command, "-")
+	for i, piece := range pieces {
+		pieces[i] = strings.ReplaceAll(piece, "_", "-")
+	}
+	return pieces
+}
+
+// Returns the unique strings in `in`, preserving order.
+//
+// Order preservation is important for search paths: a higher priority plugin
+// (path search wise) will shadow a lower priority one.
+func uniqueTrimmedStrs(in []string) []string {
+	seen := make(map[string]bool)
+	keep := in[:0]
+	for i := 0; i < len(in); i++ {
+		path := in[i]
+		path = strings.TrimSpace(path)
+		if seen[path] || len(path) == 0 {
+			continue
+		}
+		keep = append(keep, path)
+	}
+	return keep
+}
```

Comment:
I think you're never modifying `seen`, i.e. there's no `seen[path] = true`

Reply:
Hrmm I thought one of the tests caught this case... I'll make sure to touch up whatever test I thought did this. You're right.

Reply:
Turns out there is a test that hits this behavior, but the result was a log line saying this would be shadowed.

Reply:
I've written a new test for this specifically.